spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java

```diff
@@ -982,6 +989,21 @@ public Object resolveArgument(MethodParameter parameter, Message<?> message) thr
 			return resolved;
 		}
 
+		private void validateAcknowledgementUsedOnlyWithManualAckMode(MethodParameter parameter, Message<?> message) {
+			Class<?> parameterType = resolveTargetClass(parameter, message);
+			if (parameterType == Acknowledgment.class) {
+				String paramName = parameter.getParameterName();
+				paramName = (paramName != null ? paramName : "Arg " + parameter.getParameterIndex());
+				BindingResult bindingResult = new BeanPropertyBindingResult(message.getPayload(), paramName);
+				bindingResult.addError(new ObjectError(paramName,
+						"Parameter of type Acknowledgement should not be resolved by this resolver.\n" +
+						"Use Acknowledgement argument only if the listener is configured to use manual acknowledgment:\n" +
+						"\tspring.kafka.consumer.auto-offset-reset=earliest\n" +
+						"\tspring.kafka.listener.ack-mode=manual\n"));
```

Comment:
First of all we don't say anything Spring Boot-specific in the code. This project is really can be used outside of any Spring Boot features.

On the other hand I don't see how this unconditional exception makes sense: you just don't check any provided `ackMode` in this context. Plus this `KafkaNullAwarePayloadArgumentResolver` is really used to replace a regular `PayloadArgumentResolver`, so still not clear why it must reject the `Acknowledgement` param and its argument?

What you are trying to achieve is really looks like a single configuration phase validation - not against every single consumer call... Probably something like `HandlerAdapter` or somewhere else in the `KafkaListenerEndpoint` hierarchy should be better place to check expected params and configured options before any real Kafka records processing.

Reply:
Hi @artembilan , 

Thank you for taking the time to review this PR.
The details of the message may be inappropriate at this location but a validation that the payload resolver isn't trying to resolve an argument of type `Acknowledgement` does seem correct to me. If I understand the resolver purpose correctly, that shouldn't happen. Do you think I should completely remove this exception or just change the text?

I will try to look at the locations you suggested to add a validation that `Acknowledgement` argument is not used when `ackMode` is not manul and free up some time to write a proper validation there. 

Thank you, Sela

Reply:
Well, this resolver is not going to deal with `Acknowledgment` at all since it cannot be supplied in the `message.payload`. So, such a validation in this resolver is misleading. Plus we really don't talk here about actual message against particular Kafka record, but more about misconfiguration where non-manual ackMode cannot be used with `Acknowledgement` param. That's all: nothing more should be involved.